### ISSUE TITLE
Memory error at stopStream/closeStream

### ISSUE THREAD

**Issue by TpnyF** [thestk/rtaudio#265]
There are problems with stopStream/closeStream - it's really a rule.

I tried the versions: 4.1.2 , 5.0.0 , 5.1.0 They all give memory error when I close or stop the stream. Is there a problem with my code?

If give an error to such a simple application, and even on your example, I wonder how it behaves in complex applications.

`````
#include <windows.h>
#include "RtAudio.h"
#include <iostream>
#include <cstdlib>

enum Menu {
	id_play = 0,
	id_stop
};
typedef signed short MY_TYPE;
#define FORMAT RTAUDIO_SINT16
#define SCALE  32767.0

#define SLEEP( milliseconds ) Sleep( (DWORD) milliseconds ) 
#define BASE_RATE 0.005
#define TIME   1.0

HINSTANCE mHInstance;

void errorCallback( RtAudioError::Type type, const std::string &errorText )
{
  // This example error handling function does exactly the same thing
  // as the embedded RtAudio::error() function.
  std::cout << "in errorCallback" << std::endl;
  if ( type == RtAudioError::WARNING )
    std::cerr << '\n' << errorText << "\n\n";
  else if ( type != RtAudioError::WARNING )
    throw( RtAudioError( errorText, type ) );
}

unsigned int channels;
RtAudio::StreamOptions options;
unsigned int frameCounter = 0;
bool checkCount = false;
unsigned int nFrames = 0;
const unsigned int callbackReturnValue = 1;
unsigned int bufferFrames, fs, device = 0, offset = 0;
double *data = (double *) calloc( channels, sizeof( double ) );


RtAudio dac;

int saw( void *outputBuffer, void * /*inputBuffer*/, unsigned int nBufferFrames,
         double /*streamTime*/, RtAudioStreamStatus status, void *data )
{
  unsigned int i, j;
  extern unsigned int channels;
  MY_TYPE *buffer = (MY_TYPE *) outputBuffer;
  double *lastValues = (double *) data;

  if ( status )
    std::cout << "Stream underflow detected!" << std::endl;

  double increment;
  for ( j=0; j<channels; j++ ) {
    increment = BASE_RATE * (j+1+(j*0.1));
    for ( i=0; i<nBufferFrames; i++ ) {
      *buffer++ = (MY_TYPE) (lastValues[j] * SCALE * 0.08);
      lastValues[j] += increment;
      if ( lastValues[j] >= 1.0 ) lastValues[j] -= 2.0;
    }
  }

  frameCounter += nBufferFrames;
  if ( checkCount && ( frameCounter >= nFrames ) ) return callbackReturnValue;
  return 0;
}

static LRESULT CALLBACK WindowProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam) {
    switch(msg) {
		case WM_CREATE: {
			HMENU hMenu;
			hMenu = CreateMenu();

			AppendMenu(hMenu, MF_STRING, id_play, "Play");
			AppendMenu(hMenu, MF_STRING, id_stop, "Stop");

			SetMenu(hWnd, hMenu);

			if ( dac.getDeviceCount() < 1 ) {
			OutputDebugString("No audio devices found!");
			exit( 1 );
			}
			
			channels = 2;
			fs = 44100;

			// Let RtAudio print messages to stderr.
			dac.showWarnings( true );
			  bufferFrames = 512;
			  RtAudio::StreamParameters oParams;
			  oParams.deviceId = device;
			  oParams.nChannels = channels;
			  oParams.firstChannel = offset;

			  if ( device == 0 )
				oParams.deviceId = dac.getDefaultOutputDevice();
			OutputDebugString("Ready to start");

			  options.flags = RTAUDIO_HOG_DEVICE;
			  options.flags |= RTAUDIO_SCHEDULE_REALTIME;
			#if !defined( USE_INTERLEAVED )
			  options.flags |= RTAUDIO_NONINTERLEAVED;
			#endif
			  try {
				dac.openStream( &oParams, NULL, FORMAT, fs, &bufferFrames, &saw, (void *)data, &options, &errorCallback );
				}
				catch (RtAudioError& e) {
					OutputDebugString(e.getMessage().c_str());
					if (dac.isStreamOpen()) 
						dac.closeStream();
				}



			return 0;
		}
        case WM_COMMAND:
            switch(LOWORD(wParam)) {
                case id_play:
					if (!dac.isStreamRunning()) {
						try {
							dac.startStream();
						}
						catch ( RtAudioError& e ) {
						OutputDebugString(e.getMessage().c_str());
						}
					}
                break;
                case id_stop:
					if (dac.isStreamRunning()) {
						try {
							dac.stopStream();
						}
						catch ( RtAudioError& e ) {
						OutputDebugString(e.getMessage().c_str());
						}
					}
                break;
            }
        break;
        case WM_CLOSE:
			if ( dac.isStreamOpen() ) dac.closeStream();
			//free( data );
			char className[50];
			GetClassName(hWnd, className, sizeof(className));
            DestroyWindow(hWnd);
			UnregisterClass(className, mHInstance);
        break;
        case WM_DESTROY:
            PostQuitMessage(0);
        break;
        default:
            return DefWindowProc(hWnd, msg, wParam, lParam);
    }
    return 0;
}
HRESULT Create(HINSTANCE hInstance) {
	HWND hWnd; WNDCLASSEX wCs;
	static const char* wndClassName = "WndClass";

	wCs.cbSize			= sizeof wCs;
	wCs.style			= CS_HREDRAW | CS_VREDRAW;
	wCs.lpfnWndProc		= WindowProc;
	wCs.cbClsExtra		= 0;
	wCs.cbWndExtra		= 0;
	wCs.hInstance		= hInstance;
	wCs.hIcon			= LoadIcon(hInstance, IDI_APPLICATION);
	wCs.hCursor			= LoadCursor(NULL, IDC_ARROW);
	wCs.hbrBackground	= (HBRUSH)(COLOR_WINDOW+1);
	wCs.lpszMenuName	= NULL;
	wCs.lpszClassName	= wndClassName;
	wCs.hIconSm			= LoadIcon(hInstance, IDI_APPLICATION);

	if (RegisterClassEx(&wCs)) {
		hWnd = CreateWindowEx(
			WS_EX_CLIENTEDGE,			// extended window style
			wndClassName,	// registered class name
			TEXT("SimpleWindow"),	// and window title
			WS_OVERLAPPEDWINDOW,		// window style
			CW_USEDEFAULT, CW_USEDEFAULT, 600, 350,	
			NULL, NULL, hInstance,		// handle to application instance
			NULL						// if need to take pointer
		);
	}
	if(hWnd==NULL) {
		MessageBox(NULL, "Window Creation Failed!", "Error!", MB_ICONEXCLAMATION | MB_OK);
		return E_FAIL;
	}
	ShowWindow(hWnd, SW_SHOWNORMAL);
	UpdateWindow(hWnd);
	mHInstance=hInstance;		

	return S_OK;
}
int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpszArgument, int nCmdShow) {
	int nAppResult = -1;
	if (SUCCEEDED(Create(hInstance))) {
		
		MSG	msg;
		while (GetMessage(&msg, NULL, 0, 0)) {
			TranslateMessage(&msg);
			DispatchMessage(&msg);
		}
		nAppResult = msg.wParam;
	}
	return nAppResult;
}
```

**Comment by garyscavone** [comment]
Some details of your operating system and the API you are trying to use would be crucial to addressing this issue.

**Comment by TpnyF** [comment]
:) But is clear from the code: WINAPI

Windows, test with RTAudio  versions:  4.1.2 , 5.0.0 , 5.1.0

**Comment by sonoro1234** [comment]
> :) But is clear from the code: WINAPI

RtAudio API?

**Comment by TpnyF** [comment]
Sure, Wasapi is.

More info: 
In function:  RtApiWasapi::stopStream 
. . . . . . . . . . . . . . . .
  while( stream_.state != STREAM_STOPPED ) {
    Sleep( 1 );
  }

never go out from this loop when attempt to stop stream, that's probably the problem !?

**Comment by TpnyF** [comment]
I get this error with DebugView:

RtlWerpReportException failed with status code :-1073741823. Will try to launch the process directly

Failed to create the process C:\Windows\system32\WerFault.exe

**Comment by sonoro1234** [comment]
Did you test https://github.com/thestk/rtaudio/blob/master/tests/playsaw.cpp also?

**Comment by TpnyF** [comment]
playsaw.cpp crash too...

While doing tests I found a certain solution. Simple addition of the line:

std::map<int, std::string> apiMap;

Fix the problem. I do not know why ? I do not know if it is really a safe solution, I am afraid that in certain circumstances the problem will reappear.

**Comment by sonoro1234** [comment]
You will have to wait for @MarcusTomlinson to came back.
Now we have more info!!

**Comment by desmondlr** [comment]
I was also having this issue, I was not able to reproduce the error using the code that @TpnyF shared or playsaw.cpp, but I did get a crash that matched this description in a project I have been working on as well as a (not so) minimal example that I made. This suggested to me that it was caused by a race condition so I took a look at RtApiWasapi.  If the thread is closed while the playback loop is running it can cause some issues. 

I tried my hand at making RtApiWasapi thread-safe and drafted a pull request here: #278. These changes have fixed the problem for me. I tried to keep my changes as minimal as possible, but I did have to modify RtApi::error().
